File: packages/python-sdk/qelos_sdk/ai/agents.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from ..base_sdk import BaseSDK
from ..types import QelosSDKOptions
from .chat import ChatSDK, SSEStreamProcessor


class AgentsSDK(BaseSDK):
    """AI agents: list, fetch, and agent-centric chat (mirrors TS ``AgentsSDK``)."""

    _relative_path = "/api/ai/agents"

    def __init__(self, options: QelosSDKOptions, chat_sdk: ChatSDK) -> None:
        super().__init__(options)
        self._chat = chat_sdk

    async def list(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List available AI agents."""
        qs = self.get_query_params(query)
        return await self.call_json_api(f"{self._relative_path}{qs}")

    async def get(self, agent_id: str) -> Dict[str, Any]:
        """Get a specific agent by ID."""
        return await self.call_json_api(f"{self._relative_path}/{agent_id}{self.get_query_params({})}")

    async def chat(
        self,
        agent_id: str,
        message: str,
        options: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Send a single message to an agent and get the completion response."""
        built = self._build_chat_options(message, dict(options or {}))
        return await self._chat.chat(agent_id, built)

    async def stream_chat(
        self,
        agent_id: str,
        message: str,
        options: Optional[Dict[str, Any]] = None,
    ) -> SSEStreamProcessor:
        """Stream a response from an agent for a single message."""
        built = self._build_chat_options(message, dict(options or {}))
        return await self._chat.stream(agent_id, built)

    async def chat_in_thread(
        self,
        agent_id: str,
        thread_id: str,
        message: str,
        options: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Continue a conversation in an existing thread."""
        built = self._build_chat_options(message, dict(options or {}))
        return await self._chat.chat_in_thread(agent_id, thread_id, built)

    async def stream_chat_in_thread(
        self,
        agent_id: str,
        thread_id: str,
        message: str,
        options: Optional[Dict[str, Any]] = None,
    ) -> SSEStreamProcessor:
        """Stream a response within an existing thread."""
        built = self._build_chat_options(message, dict(options or {}))
        return await self._chat.stream_in_thread(agent_id, thread_id, built)

    @staticmethod
    def _build_chat_options(message: str, options: Dict[str, Any]) -> Dict[str, Any]:
        history = options.pop("history", None) or []
        messages = [*history, {"role": "user", "content": message}]
        return {**options, "messages": messages}

File: packages/sdk/src/ai/agents.ts
import BaseSDK from '../base-sdk';
import { QelosSDKOptions } from '../types';
import ChatSDK from './chat';
import ThreadsSDK from './threads';
import {
  IAgent,
  IAgentChatOptions,
  IChatCompletionResponse,
  ISSEStreamProcessor,
  IMessage
} from '.';

export default class AgentsSDK extends BaseSDK {
  private relativePath = '/api/ai/agents';

  constructor(
    options: QelosSDKOptions,
    private chatSdk: ChatSDK,
    private threadsSdk: ThreadsSDK
  ) {
    super(options);
  }

  /**
   * List available AI agents
   */
  list(query?: {
    active?: boolean;
    kind?: string;
    limit?: number;
    page?: number;
    sort?: string;
    [key: string]: string | number | boolean;
  }): Promise<IAgent[]> {
    return this.callJsonApi<IAgent[]>(`${this.relativePath}${this.getQueryParams(query)}`);
  }

  /**
   * Get a specific agent by ID
   */
  get(agentId: string): Promise<IAgent> {
    return this.callJsonApi<IAgent>(`${this.relativePath}/${agentId}${this.getQueryParams({})}`);
  }

  /**
   * Send a single message to an agent and get the completion response.
   * No thread is persisted — use `chatInThread` to maintain conversation continuity.
   */
  chat(
    agentId: string,
    message: string,
    options: IAgentChatOptions = {}
  ): Promise<IChatCompletionResponse> {
    return this.chatSdk.chat(agentId, this.buildChatOptions(message, options));
  }

  /**
   * Stream a response from an agent for a single message.
   * Returns an iterable of parsed SSE chunks.
   */
  async streamChat(
    agentId: string,
    message: string,
    options: IAgentChatOptions = {}
  ): Promise<ISSEStreamProcessor> {
    const stream = await this.chatSdk.stream(agentId, this.buildChatOptions(message, options));
    return this.chatSdk.parseSSEStream(stream);
  }

  /**
   * Continue a conversation in an existing thread by sending a new user message.
   */
  chatInThread(
    agentId: string,
    threadId: string,
    message: string,
    options: IAgentChatOptions = {}
  ): Promise<IChatCompletionResponse> {
    return this.chatSdk.chatInThread(agentId, threadId, this.buildChatOptions(message, options));
  }

  /**
   * Stream a response from an agent within an existing thread.
   * Returns an iterable of parsed SSE chunks.
   */
  async streamChatInThread(
    agentId: string,
    threadId: string,
    message: string,
    options: IAgentChatOptions = {}
  ): Promise<ISSEStreamProcessor> {
    const stream = await this.chatSdk.streamInThread(
      agentId,
      threadId,
      this.buildChatOptions(message, options)
    );
    return this.chatSdk.parseSSEStream(stream);
  }

  private buildChatOptions(message: string, options: IAgentChatOptions) {
    const { history, ...rest } = options;
    const messages: IMessage[] = [
      ...(history || []),
      { role: 'user', content: message }
    ];
    return { ...rest, messages };
  }
}
